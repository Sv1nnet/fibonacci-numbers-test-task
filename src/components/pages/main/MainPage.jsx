import React, { useState, useEffect } from 'react';
import NumbersInput from '@components/ui/inputs/numbersInput/NumbersInput';
import Result from '@components/result/Result';
import RedirectButton from '@components/ui/buttons/redirectButton/RedirectButton';
import { useSendNumber } from '@src/hooks/api/apiHooks';

const MAX_NUMBER = 1476;

const MainPage = () => {
  const [number, setNumber] = useState(null);
  const [errorMessage, setErrorMessage] = useState('');
  const [{ data, loading, error }, getResult] = useSendNumber();

  const sendCalcRequest = () => {
    if (number !== null && number >= 0) {
      if (number < MAX_NUMBER) {
        setErrorMessage('');
        getResult(number);
      } else {
        setErrorMessage(`Number can't be greater than ${MAX_NUMBER}`);
      }
    }
  };

  // This handler only executes if target.value is a number or empty string
  const onChange = (e) => {
    const { target } = e;

    if (target.value !== '') setNumber(parseInt(target.value, 10));
    else setNumber(null);
  };

  const onKeyUp = (e) => {
    if (e.key === 'Enter') sendCalcRequest();
  };

  useEffect(() => {
    if (error) setErrorMessage(error.message || error.statusText || '');
  }, [error]);

  return (
    <div>
      {
        errorMessage || error
          ? <Result error result={errorMessage} />
          : <Result result={loading && !error ? 'Calculating...' : data} />
      }
      <NumbersInput onKeyUp={onKeyUp} onChange={onChange} disabled={loading} />
      <p className="center-align">Press Enter to calculate</p>
      <RedirectButton to="history" text="View previous results" />
    </div>
  );
};

export default MainPage;
