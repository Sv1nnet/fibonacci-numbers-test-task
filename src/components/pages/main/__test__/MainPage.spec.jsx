import React from 'react';
import { it, describe, expect } from '@jest/globals';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import MainPage from '../MainPage';

configure({ adapter: new Adapter() });

describe('Test MainPage', () => {
  it('Should match MainPage snapshot', () => {
    const component = shallow(<MainPage />);

    expect(toJson(component)).toMatchSnapshot();
  });
});
