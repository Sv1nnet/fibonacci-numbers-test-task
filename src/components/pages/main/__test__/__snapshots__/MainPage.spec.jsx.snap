// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MainPage Should match MainPage snapshot 1`] = `
<div>
  <Result
    error={false}
    result={null}
  />
  <NumbersInput
    disabled={false}
    onChange={[Function]}
    onKeyUp={[Function]}
  />
  <p
    className="center-align"
  >
    Press Enter to calculate
  </p>
  <RedirectButton
    className=""
    text="View previous results"
    to="history"
  />
</div>
`;
